### DISCUSSION TITLE
How can i return User of different structure from authorize?

### DISCUSSION THREAD

**Discussion by typedashutosh** [nextauthjs/next-auth#1599] [Help] [answered]
In [...nextauth].ts, I've used credentials for signing in a client. after signing in, the `authorize` method should return a `User` object with `id` `email` and `image` . 

**But** I want to return a custom `UserObject` with `_id`, `firstname`, `lastname` and `username` as parameters.

As soon as I does this, typescript start complaining this -->
`Type 'Promise<UserObject | null>' is not assignable to type 'Promise<User | null>'.`

The only way I could think of accomplishing this is by making changes in `node_modules` which doesn't seem to be good idea.

I also couldn't find any documentation or examples doing this, if found, weren't helpful or hard to understand.

```
dependencies: {
    "@types/next-auth": "^3.7.4",
    "next-auth": "^3.11.2",
    "typescript": "^4.2.3"
}
```

**Comment by rbatistajs** [comment, marked as answer]
I solved it by adding types.d.ts in my project. 

extends interface "User" the of module "next-auth"
```
import 'next-auth'

declare module 'next-auth' {
  export interface User {
    _id?: string
    firstname?: string
    lastname?: string
    username?: string
  }
}
```

example extends other interface
```
/* eslint @typescript-eslint/no-empty-interface: "off" */
import 'next-auth'
import { UserObject } from './my-code/user'

declare module 'next-auth' {
  export interface User extends UserObject  {}
}
```



![image](https://user-images.githubusercontent.com/2729379/113954365-0f536180-97f0-11eb-9af0-8781120324d6.png)


=============


_Would be cool to do something like this in next-auth :/_
![image](https://user-images.githubusercontent.com/2729379/113954179-c00d3100-97ef-11eb-9044-226f6e287752.png)

**Reply by balazsorban44** [reply]
Working on it! :) #1649 has started this process, and I had the same idea at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51926

**Reply by typedashutosh** [reply]
This method of explicitly defining types in separate file does work! Also I'd to include this:
```
declare module 'next-auth/jwt' {
  export interface JWT extends Record<string, unknown> {
    _id: string
    firstname: string
    lastname: string
    username: string
  }
``` 
so that i could return correct session to client with `useSession()`.
Still, I'm getting these useless properties:

![image](https://user-images.githubusercontent.com/68853189/114130425-2c864e00-991e-11eb-806a-ea45f3007c74.png)

In your example, its also present there.

**Question:**
Is there a way to override default `User` and `JWT` types declared in `@types/next-auth`?

**Reply by balazsorban44** [reply]
I believe currently there is not. But I did consider type augmentation like you did here, and it would work even better than generics, especially for jwt and session.

cc @lluia please remind me of this next time we discuss this topic. 😊

**Reply by typedashutosh** [reply]
I've an Idea!
Can we have a `user` constructor, where we can specify  user object, just like mongoose  schema.
Example of mongoose schema:
```
const UserSchema = new Schema<IUser>(
  {
    firstname: { type: String, required: [true, 'Firstname is required'] },
    lastname: String,
    username: {
      type: String,
      unique: true,
      required: [true, 'Please provide a username']
    },
    password: { type: String, required: [true, 'Please provide a Password'] },
    blogs: { type: Array, required: false }
  }
)
```

We can have limited options for `keys` in `user` object like `required`.

What you say?

**Reply by balazsorban44** [reply]
will be addressing this through https://next-auth-git-feat-deduplicate-types-nextauthjs.vercel.app/getting-started/typescript
you will be able to augment `User`

This will soon be built-in you an follow #1690
